### ISSUE TITLE
New First Screen - two freely moving objects

### ISSUE THREAD

**Issue by emily-phet** [phetsims/proportion#7]
Let's move the current prototype screen to the second screen. The new first screen would have, as DA mentioned yesterday, two freely moving points. Maybe two small spheres?

@zepumph can you comment on any immediate implementation issues you see with this. I recognize that freely moving objects will introduce downstream challenges for interaction, description, etc., but if you know of any immediate issues beyond the standard freely-moving-objects-are-difficult issues, please point them out here. 

Design notes - Let's try first with having the reset all button, and radio buttons just free floating. The whole sim background color changes to indicate "success" or not. Will need to be more pastel variations than the bar graph colors. Let's start with white background = not at ratio, and a medium-dark pastel green = at ratio ([maybe this green](https://colorhunt.co/palette/177869) - second from top). 

Future considerations (if we move forward with public publication of this screen)
- input options - streamlining alternative access to the objects.
- all the ways we can cue proportion success - probably cannot rely on color change (visual) and sound (auditory) alone. One thought - changing object shape as well, for a second visual cueing? Floating indicator that changes as success is reached (like a "close, close, close, checkmark success!" visual indicator. I recognize that this may be counter to goals of having the learner attend to the changing visual height difference, to their fingers, etc., but just putting  it out there for consideration for some learners. Could have this sort of secondary visual cueing an option that could be enabled from the options menu, not on by default.

**Comment by emily-phet** [comment]
Additional considerations:

- Let's have an option to switch the freely moving points from small spheres, to cross hairs (radio button options).
- Model update - *any* way the height ratio = the desired ratio (e.g., 1:2) results in "success" feedback (color change, etc.). Doesn't matter whether left or right object is higher or lower, as long as the height ratio is the desired ratio.

**Comment by zepumph** [comment]
I have implemented the second (actually first) screen in the above commits.

Here is the latest dev version: https://phet-dev.colorado.edu/html/proportion/1.0.0-dev.3/phet/proportion_en_phet.html

My goal was largely to implement things as simply as possible, and to wait for more design feedback before honing things in. I don't even feel like I have a good name for the screen just yet. I went with "Freely Moving Objects" as the title, and in the code `freeObjects`, and I called each movable proportion piece a "marker." I'm not sold on those words, but I thought it would be good so that we can initially be talking about the same things. 

>Let's have an option to switch the freely moving points from small spheres, to cross hairs (radio button options).

Implemented with radio buttons

>Model update - any way the height ratio = the desired ratio (e.g., 1:2) results in "success" feedback (color change, etc.). Doesn't matter whether left or right object is higher or lower, as long as the height ratio is the desired ratio.

I have implemented this in both screens. Because of this, I got rid of "Mystery 3" which was actually `4:1`, because order no longer matters in it. 

>Will need to be more pastel variations than the bar graph colors.

From refactoring, I have made both screens use this new green color from the above comment.

>input options - streamlining alternative access to the objects.

Right now I have only implemented a basic keyboard drag listener for each "marker." Just like in the other screen. I look forward to refining the interaction more.

>all the ways we can cue proportion success - probably cannot rely on color change (visual) and sound (auditory) alone. One thought - changing object shape as well, for a second visual cueing? Floating indicator that changes as success is reached (like a "close, close, close, checkmark success!" visual indicator. I recognize that this may be counter to goals of having the learner attend to the changing visual height difference, to their fingers, etc., but just putting it out there for consideration for some learners. Could have this sort of secondary visual cueing an option that could be enabled from the options menu, not on by default.

I didn't implement this

Here are some further considerations that I would want feedback on before spending time on:
* dilate pointer areas for the markers?
* What should we do about cue arrows for this screen?
* Should we do something to mark the bottom. It is implemented with what is called the "layout bounds" of a sim, which doesn't necessarily go all the way down to the navigation bar. The layout bounds are drawn if you add the `?dev` query parameter (if you are curious). This could be alright because the green background only covers the layout bounds, so in a weak way it is already defined.

Tagging @BLFiedler because he had good feedback last time.

**Comment by DorAbrahamson** [comment]
Emily, let's revisit the tradeoffs of allowing an a:b ratio to be green both when left < right (a:b) AND left > right (b:a). I have always preferred to give green only for the a:b, because a:b =\= b:a. (That said, on the other project, where two students are facing each other, we would need to exercise flexibility). This is my First Post Ever on GitHub, so Hello World and apologies if I put this in the wrong place.

**Comment by emily-phet** [comment]
Welcome to GitHub @DorAbrahamson!

Thanks for bringing this up. This is a good potential "fork in the road" we'll want to sort out soon. It may be particularly significant for those working on interactive description, and maybe sound. 

If it's something we'll need to investigate empirically, we can pick the most flexible option and proceed, but we'll want the team to know that so they can design with a potential future change in direction in mind.

**Comment by BLFiedler** [comment]
First two thoughts are from a cursor/visual/touchscreen perspective, third more universal, fourth in response to Dor & EM:

1. @zepumph There's a bit of "momentum" to dragging the object, which I'm a little uncertain of since it pulls the visual away from your body movement a little when moving quickly (I notice this pretty strongly using my hands and a touch screen). I'd like to hear others' thoughts on this.

2. It might be worth considering the optimal size of the objects - slightly bigger or asymmetric objects would let me see them under my fingers. However, perhaps not seeing the objects and only relating to hand movement

3. Likely not a priority, but given the emphasis of these screens on the Y-dimension (vertical), having the sim reshape to portrait mode on a touchscreen would be nice to give a bit more room to play with in that dimension.

4. @DorAbrahamson I agree this is especially confusing in the absence of further scaffolding (like the grids/numbers). Would you think that a (young) learner might relate to say fractions and think them reversible? Or is there a more primitive proportional understanding at stake if we equate both sides?

We may need some scaffolding on (visual or auditory) display to indicate the two objects are not the same if we make the sim asymmetric.

**Comment by DorAbrahamson** [comment]
Hi! Responding to Brett, see below:

> On Apr 7, 2020, at 7:51 AM, Brett Fiedler <notifications@github.com> wrote:
> 
> 
> First two thoughts are from a cursor/visual/touchscreen perspective, third more universal, fourth in response to Dor & EM:
> 
> @zepumph <https://github.com/zepumph> There's a bit of "momentum" to dragging the object, which I'm a little uncertain of since it pulls the visual away from your body movement a little when moving quickly (I notice this pretty strongly using my hands and a touch screen). I'd like to hear others' thoughts on this.
> 
> It might be worth considering the optimal size of the objects - slightly bigger or asymmetric objects would let me see them under my fingers. However, perhaps not seeing the objects and only relating to hand movement
> 
> Likely not a priority, but given the emphasis of these screens on the Y-dimension (vertical), having the sim reshape to portrait mode on a touchscreen would be nice to give a bit more room to play with in that dimension.
> 
> @DorAbrahamson <https://github.com/DorAbrahamson> I agree this is especially confusing in the absence of further scaffolding (like the grids/numbers). Would you think that a (young) learner might relate to say fractions and think them reversible? Or is there a more primitive proportional understanding at stake if we equate both sides?
> 
1. Momentum would be a nice, realistic effect. Still, I worry about confusion over the “where” of my finger. I think finger-to-cursor tightness is less ambiguous.

2. Thinking ahead to the appearance of the grid, I’d want the player to see the cursor’s precise location on the grid. The crosshairs of the “plus” cursor enable acuity in calibrating cursor location. 

3. I like this idea of leveraging the mobile app native affordances to enable both vertical and horizontal. We have data of kids who talked through the a:b problem by citing two cards driving on the freeway at different speeds. 

4. I think of the sim interactions as the Ground 0 of mathematical meaning. I worry that students would come out of the sim thinking that a:b = b:a , because that’s what the green-either-way would suggest. This frame of mind could carry over to numbers, where kids will think that 2:3 = 6:4 because they’re both green. So my consideration is pedagogical, with an orientation toward the concepts. 

* Tradeoff 1: having both a:b and b:a green means that, on average over many users, you’ll happened to find your first green faster, which is good for retention (keeping people un-frustrated). 
*Tradeoff 2: Arguably, there is some conceptual gain in noticing both a:b and b:a, because then you are likely to ask yourself “In what sense are these how things the same such that they are both green?"

-Dor.

**Comment by BLFiedler** [comment]
@zepumph Is the tolerance multiplicative? We might consider a constant tolerance. It seems I have more wiggle room at larger values than smaller values.

**Comment by DorAbrahamson** [comment]
(Tolerance has always been a sticky issue for us. On the one hand, an absolute tolerance is impracticable at the lower values, because, then, just about everything is green. So, in the lower values, a universal absolute margin would be too large. On the other hand, a multiplicative margin would be too small in the lower values, to the point where it would be down to a single pixel! And so, in the past we have done some combination of these that transitions from some minimal absolute margin in the lower values and then transitioning into an absolute margin for the greater values. It was not elegant, but it worked. I don’t think that students ever felt a hitch or hiccup as their hands moved along from lower to higher values.)

**Comment by BLFiedler** [comment]
Great insight. A piecewise function is a sound strategy for our first pass. I suspect you may have an idea for each proportion when the "feel" is right to switch between either strategy.

**Comment by DorAbrahamson** [comment]
Well, one might wish to begin by asking what is a minimal acceptable distance between “left” and “right.” That could serve us as a basis for evaluating when to switch per each ratio. That said, given that we hope to be multi-platform (yes?), such as various multi-touch PDA, tablet, and desktop, this minimum would need to vary by host platform… I wouldn't know if that is technically feasible.

**Comment by zepumph** [comment]
In regards to tolerance: 
Currently the tolerance is constant, and always the value in the NumberControl. This value itself is a bit arbitrary, but it is the same unit as the values of the left/right objects. Currently the first screen's objects can range in height from 0 (bottom) to ~.6 (top). This is totally arbitrary based on how the view elements are set up and scaled from the model. Thus a tolerance of .05 is about ~8% of the vertical draggable real estate. 

In regards to multi-device support:
The "layout bounds" is what we call the space in the sim in which you can drag the objects around. This rectangle is defined generally, and is often consistent across sims. It scales based on the device screen dimensions and will always have the same aspect ratio. From this note that we can create values for tolerance etc in the model that will behave very similarly across devices since they will all behave in a way that is constrained within the layout bounds of the sim. 

In regards to reciprocal support for proportions:
I am no expert! Either way is just as easy to implement; just let me know.

**Comment by BLFiedler** [comment]
@zepumph Sorry, I may be misunderstanding. There is a difference in the distance I move the right circle/crosshair and have a green screen when the left circle/crosshair is either very low or in the middle of the screen. It's exaggerated if I raise the tolerance to say "0.20", and when the left circle/crosshair is in the middle of the screen there is a large distance I can move the right crosshair while maintaining a green screen than if the left is very low. This tells me the tolerance is actually dependent on the height of the lower circle/crosshair?

**Comment by zepumph** [comment]
Ahh I see that! You are right.

**Comment by DorAbrahamson** [comment]
My hunch would be to try out various versions, making sure we strike a balance between, on the on hand, having some minimal differentiation at the low values near the bottom of the screen where it’s crowded and, on the other hand, having a more-or-less uniform margin higher up. Of course, arguable, if the ratio is 9:10…, then even higher up on the screen we cannot have large margins (bc/ then 10:10 might be greenish). So perhaps the algorithm should be linked to the a/b relation.

**Comment by BLFiedler** [comment]
@zepumph This issue has gone on a bit of a tangent. I've opened up a new issue (#14) to discuss the tolerance issue. 

I've identified the following unanswered/unimplemented questions from earlier:

Your questions:
>Here are some further considerations that I would want feedback on before spending time on:
> - dilate pointer areas for the markers?
> - What should we do about cue arrows for this screen?
> - Should we do something to mark the bottom. It is implemented with what is called the "layout bounds" of a sim, which doesn't necessarily go all the way down to the navigation bar. The layout bounds are drawn if you add the ?dev query parameter (if you are curious). This could be alright because the green background only covers the layout bounds, so in a weak way it is already defined.

and 

> all the ways we can cue proportion success - probably cannot rely on color change (visual) and sound (auditory) alone. One thought - changing object shape as well, for a second visual cueing? Floating indicator that changes as success is reached (like a "close, close, close, checkmark success!" visual indicator. I recognize that this may be counter to goals of having the learner attend to the changing visual height difference, to their fingers, etc., but just putting it out there for consideration for some learners. Could have this sort of secondary visual cueing an option that could be enabled from the options menu, not on by default.

Are any of these worthy of new issues @zepumph? 

Or @emily-phet or @DorAbrahamson do you have any thoughts?

re: cue arrows - Perhaps similar to other free moving sims like the 4 direction arrows on the magnet in Faraday's Law?
![image](https://user-images.githubusercontent.com/47331395/79135486-99e3bb00-7d6c-11ea-98d0-092bf005d766.png)

**Comment by zepumph** [comment]
>re: cue arrows - Perhaps similar to other free moving sims like the 4 direction arrows on the magnet in Faraday's Law?

moved to https://github.com/phetsims/proportion/issues/18

I moved most questions to separate issues. @BLFiedler anything else for this issue?

**Comment by BLFiedler** [comment]
I think we're good! Closing now as we move to specific discussions about Screen 1.
